---
title: "DEseqOnSpruce"
output: html_document
date: "2025-11-24"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = 
                       "~/projects/eco_genomics_2025/finalproject/")
```

### libraries and data

```{r}
library(DESeq2)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(vsn)  
library("pheatmap")
library("vsn")

# import data
setwd("~/projects/eco_genomics_2025")
countsTable <- read.csv("final_project/mydata/counts_matrix.csv", 
                        header=TRUE, row.names=1)
countsTableRound  <- round(countsTable)
samples <- colnames(countsTableRound)
keep <- !grepl("BLANK", samples)
countsTableRound <- countsTableRound[, keep]
samples <- colnames(countsTableRound)

createMeta <- function(samples){
  parts <- unlist(strsplit(samples, "_"))
  sourcepop <- paste(parts[1], parts[2], sep = "_")
  treatment <- parts[parts %in% c("C", "H", "D")]
  treatmentmap <- c(
    C = "Control",
    H = "Heat",
    D = "Drought"
  )
  treatments <- unname(treatmentmap[treatment])
  treatment_string <- paste(unique(treatments), collapse = "; ")
  lasttreatindex <- max(which(
    parts %in% c("C", "H", "D")))
  day <- as.integer(parts[lasttreatindex + 1])
  data.frame(source_population = sourcepop,
             treatment = treatment_string,
             day_harvest = day,
             stringsAsFactors = FALSE)
}
meta_list <- lapply(samples, createMeta)
metadataSpruceRNA <- bind_rows(meta_list) 
metadataSpruceRNA$samples <- samples
rownames(metadataSpruceRNA) <- metadataSpruceRNA$samples
metadataSpruceRNA <- metadataSpruceRNA %>%
  mutate(source_climate = case_when(
    source_population %in% c("ASC_06","BRU_05",
                             "ESC_01", "XBM_07",
                             "NOR_02") ~ "HotDry",
    source_population %in% c("CAM_02", "JAY_02",
                             "KAN_04", "LOL_02",
                             "MMF_13") ~ "CoolWet"
  ))

file <- "metadata_spruce.csv"
write.csv(metadataSpruceRNA, file = file, row.names = FALSE)
```

### data distribution

```{r}
# Let's see how many reads we have from each sample
colSums(countsTableRound)
mean(colSums(countsTableRound))

barplot(colSums(countsTableRound), names.arg=colnames(countsTableRound),cex.names=0.5, las=3,ylim=c(0,2500000)) #change dims
abline(h=mean(colSums(countsTableRound)), col="blue", lwd=2)

# the average number of counts per gene
rowSums(countsTableRound)
mean(rowSums(countsTableRound)) # 
median(rowSums(countsTableRound)) #

apply(countsTableRound,2,mean) # 2 in the apply function does the action across columns
apply(countsTableRound,1,mean) # 1 in the apply function does the action across rows
hist(apply(countsTableRound,1,mean),xlim=c(0,14), ylim=c(0,60000),breaks=1000000)
```

### DEseq2

```{r}
dds_sourceVday <- DESeqDataSetFromMatrix(countData = countsTableRound,
                              colData=metadataSpruceRNA, 
                              design= ~ source_climate + day_harvest)
dim(dds_sourceVday)
dds_sourceVday <- dds_sourceVday[rowSums(counts(dds_sourceVday) >= 15) >= 18,]
dds_sourceVday <- DESeq(dds_sourceVday)
resultsNames(dds_sourceVday)
# source climate versus treatment
dds_sourceVtreatment <- DESeqDataSetFromMatrix(countData = countsTableRound,
                              colData=metadataSpruceRNA, 
                              design= ~ source_climate + treatment)
dim(dds_sourceVtreatment)
dds_sourceVtreatment <- dds_sourceVtreatment[rowSums(counts(dds_sourceVtreatment) >= 15) >= 18,]
dds_sourceVtreatment <- DESeq(dds_sourceVtreatment)
resultsNames(dds_sourceVtreatment)
```

### sample clustering and visualization

```{r}
# this gives log2(n + 1)
ntd <- normTransform(dds_sourceVday)
meanSdPlot(assay(ntd))

# Variance stabilizing transformation
vsd <- vst(dds_sourceVday, blind=FALSE)
meanSdPlot(assay(vsd))
```

### heatmap

```{r}
sampleDists <- dist(t(assay(vsd)))

library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$line, vsd$generation, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Purples")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```


### cluster tree

```{r}
sampleTree <- hclust(dist(sampleDists), method="average")
# plot
plot(sampleTree, main="Sample clustering to detect outliers", sub="", xlab="",cex.lab=1.5, cex.axis=1.5, cex.main=2)
```

### PCA

```{r}
# first transform the data for plotting using variance stabilization
vsd <- vst(dds_sourceVday, blind=FALSE)

pcaData <- plotPCA(vsd,
                   intgroup=c("source_climate","day_harvest"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData,"percentVar"))

ggplot(pcaData, aes(PC1, PC2, color=day_harvest,
                    shape=source_climate)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed() +
  theme_linedraw()
```

### PCA2

```{r}
ggplot(pcaData, aes(PC1, PC2)) +
  geom_point(size=5, stroke = 1, aes(fill=source_climate, shape=treatment, alpha = day_harvest)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  #scale_shape_manual(values=c(21,22,23,24), labels = c("G1", "G2","G3", "G4"))+
  #scale_shape_manual(values=c(21,22,23,24) -> circle, square, diamond, triangle
  scale_fill_manual(values=c('#6699CC',"#CC3333"), labels = c("Control", "Treatment"))+
  #scale_alpha_manual(values = c(1.0, 0.8, 0.7, 0.3), guide = "none") +
  theme_linedraw() +
  theme(legend.position = "none") +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 3))+
  theme(text = element_text(size = 20)) +
  theme(legend.title = element_blank()) +
  theme_linedraw()
```
  - there doesn't look to be much clustering in the PCA based on any of the variables 
  - lets try isolating specific contrasts
  
### source_climate versus treatment
```{r}
metadataSpruceRNA$day_harvest <- as.character(metadataSpruceRNA$day_harvest)
dds_sourceVday$source_climate <- relevel(dds_sourceVday$source_climate, ref = "HotDry")
dds_sourceVtreatment$treatment <- relevel(dds_sourceVtreatment$treatment, ref = "Control")
dds_sourceVday$day_harvest <- relevel(dds_sourceVday$day_harvest, ref = "0")


dds_sourceVday$group <- factor(paste0(dds_sourceVday$source_climate, dds_sourceVday$day_harvest))
design(dds_sourceVday) <- ~ group
dds_sourceVday <- DESeq(dds_sourceVday)
resultsNames(dds_sourceVday)
# source V treatment
dds_sourceVtreatment$group <- factor(paste0(dds_sourceVday$source_climate, dds_sourceVday$treatment))
design(dds_sourceVtreatment) <- ~ group
dds_sourceVtreatment <- DESeq(dds_sourceVtreatment)
resultsNames(dds_sourceVtreatment)

#source climates compared, at all time points
coolwetVhotdry_t0 <- results(dds_sourceVday,
                             contrast=c("group","CoolWet0","HotDry0"), 
                             alpha = 0.05)
coolwetVhotdry_t5 <- results(dds_sourceVday,
                             contrast=c("group","CoolWet5","HotDry5"), 
                             alpha = 0.05)
coolwetVhotdry_t10 <- results(dds_sourceVday,
                              contrast=c("group","CoolWet10","HotDry10"), 
                              alpha = 0.05)
#temporal
hotdry_0v5 <- results(dds_sourceVday, contrast=c("group","HotDry0","HotDry5"), alpha = 0.05)
hotdry_0v10 <- results(dds_sourceVday, contrast=c("group","HotDry0","HotDry10"), alpha = 0.05)
hotdry_5v10 <- results(dds_sourceVday, contrast=c("group","HotDry5","HotDry10"), alpha = 0.05)
coolwet_0v5 <- results(dds_sourceVday, contrast=c("group","CoolWet0","CoolWet5"), alpha = 0.05)
coolwet_0v10 <- results(dds_sourceVday, contrast=c("group","CoolWet0","CoolWet10"), alpha = 0.05)
coolwet_5v10 <- results(dds_sourceVday, contrast=c("group","CoolWet5","CoolWet10"), alpha = 0.05)

#treatments
hotdry_CvH <- results(dds_sourceVtreatment,
                      contrast=c("group","HotDryControl","HotDryHeat"), 
                      alpha = 0.05)
hotdry_CvDH <- results(dds_sourceVtreatment,
                      contrast=c("group","HotDryControl","HotDryDrought; Heat"), 
                      alpha = 0.05)
hotdry_HvDH <- results(dds_sourceVtreatment,
                      contrast=c("group","HotDryHeat","HotDryDrought; Heat"), 
                      alpha = 0.05)
coolwet_CvH <- results(dds_sourceVtreatment,
                      contrast=c("group","CoolWetControl","CoolWetHeat"), 
                      alpha = 0.05)
coolwet_CvDH <- results(dds_sourceVtreatment,
                      contrast=c("group","CoolWetControl","CoolWetDrought; Heat"), 
                      alpha = 0.05)
coolwet_HvDH <- results(dds_sourceVtreatment,
                      contrast=c("group","CoolWetHeat","CoolWetDrought; Heat"), 
                      alpha = 0.05)
```

### plot above results

```{r}
plotMA(coolwetVhotdry_t0, ylim=c(-5,5))
plotMA(coolwetVhotdry_t5, ylim=c(-5,5))
plotMA(coolwetVhotdry_t10, ylim=c(-5,5))
plotMA(hotdry_0v5, ylim=c(-5,5))
plotMA(hotdry_0v10, ylim=c(-5,5))
plotMA(hotdry_5v10, ylim=c(-5,5))
plotMA(coolwet_0v5, ylim=c(-5,5))
plotMA(coolwet_0v10, ylim=c(-5,5))
plotMA(coolwet_5v10, ylim=c(-5,5))
plotMA(hotdry_CvH, ylim=c(-5,5))
plotMA(hotdry_CvDH, ylim=c(-5,5))
plotMA(hotdry_HvDH, ylim=c(-5,5))
plotMA(coolwet_CvH, ylim=c(-5,5))
plotMA(coolwet_CvDH, ylim=c(-5,5))
plotMA(coolwet_HvDH, ylim=c(-5,5))
```

### volcano plots
```{r}
# Make a dataframe
vol <- function(data, title){
res_df <- as.data.frame(data)
res_df <- res_df[!is.na(res_df$padj), ]
res_df$significance <- "Not Sig"
res_df$significance[res_df$padj < 0.05 & res_df$log2FoldChange > 0] <- "Up"
res_df$significance[res_df$padj < 0.05 & res_df$log2FoldChange < 0] <- "Down"
ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), color = significance)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = c("Up" = "red", "Down" = "blue", "Not Sig" = "grey")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +
  labs(
    title = title,
    x = expression(log[2]~Fold~Change),
    y = expression(-log[10]~Adjusted~p~value)
  ) +
  theme_bw(base_size = 16) +
  theme(legend.title = element_blank())
}

vol(coolwetVhotdry_t0, "CWvHT t0")
vol(coolwetVhotdry_t5, "CWvHT t5")
vol(coolwetVhotdry_t10, "CWvHT t10")
vol(hotdry_0v5, "0v5 HD")
vol(hotdry_0v10, "0v10 HD")
vol(hotdry_5v10, "5v10 HD")
vol(coolwet_0v5, "0v5 CW")
vol(coolwet_0v10, "0v10 CW")
vol(coolwet_5v10, "5v10 CW")
vol(hotdry_CvH, "CvH HD")
vol(hotdry_CvDH, "CvDH HD")
vol(hotdry_HvDH, "HvDH HD")
vol(coolwet_CvH, "CvH CW")
vol(coolwet_CvDH, "CvDH CW")
vol(coolwet_HvDH, "HvDH CW")

```

