# Transcriptomics Notebook

##### **Fall 2025 Ecological Genomics - [Class website & tutorials.](https://pespenilab.github.io/Ecological-Genomics/)**

##### **Author: G. Rei Jia**

This will keep my notes on transcriptomics coding sessions!

#### October 7, 2025 - [Clean and visualize reads](https://pespenilab.github.io/Ecological-Genomics/Fall2025/tutorials/EcoGen2025_Transcriptomics2_VisualizeCleanVisualize.html)

Today, we started the Transcriptomics module of the course!!!!

We learned about the motivation and process of asking questions and analyzing RNAseq data.

We then accessed raw data files, which can be found here:

-   `/gpfs1/cl/ecogen/pbio6800/Transcriptomics/RawData`

We then chose our samples (mine is C1) and edited the bash script, titled `fastp_tonsa.sh`. We ran into some issues with the script and will edit it next class.

#### October 9, 2025 - [Clean and visualize reads](https://pespenilab.github.io/Ecological-Genomics/Fall2025/tutorials/EcoGen2025_Transcriptomics2_VisualizeCleanVisualize.html)

We began today's coding session by editing and running the `fastp_tonsa.sh` bash script, which generated fastq reports and is housed in my `myresults/fastp_reports` folder.

We then copied the script `salmon_quant.sh` from `/gpfs1/cl/ecogen/pbio6800/Transcriptomics/scripts` into our own `myscripts/` folder and edited it and customized the script based on our repositories and sent in the mapping script.
