---
title: "DESeq2_tonsa_multigen"
output: html_document
date: "2025-10-14"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/projects/eco_genomics_2025/transcriptomics/mydata")
```

Import the libraries that we're likely to need in this session
```{r message=FALSE, warning=FALSE}
library(DESeq2)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(vsn)  
library("pheatmap")
library("vsn")
```

Import data:
```{r} 
# Import the counts table
countsTable <- read.table("counts_table.txt", header=TRUE, row.names=1)
head(countsTable)
dim(countsTable)
```

```{r}
countsTableRound <- round(countsTable) # bc DESeq2 doesn't like decimals (and Salmon outputs data with decimals)
head(countsTableRound)
```

```{r}
#import the sample description table
conds <- read.delim("metadata.txt", header=TRUE, stringsAsFactors = TRUE, row.names=1)
head(conds)
```

Explore data distributions:
```{r}
# Let's see how many reads we have from each sample
colSums(countsTableRound)
mean(colSums(countsTableRound)) #note that we are below the general rule of thumb recommendation of 20M reads per sample. We're at ~12M reads/sample.
```

Visualizing variation across samples:
```{r}
barplot(colSums(countsTableRound), names.arg=colnames(countsTableRound),cex.names=0.5, las=3,ylim=c(0,21000000)) #change dims
abline(h=mean(colSums(countsTableRound)), col="blue", lwd=2)
```

Visualizing variance across genes:

- Despite lots of zeros across the 119,000 transcripts, we still have an average of ~2500 reads mapping per transcript.
```{r}
# the average number of counts per gene
mean(rowSums(countsTableRound))
median(rowSums(countsTableRound))
```

```{r message=FALSE, warning=FALSE}
apply(countsTableRound,2,mean) # 2 in the apply function does the action across columns
apply(countsTableRound,1,mean) # 1 in the apply function does the action across rows
```

```{r}
hist(apply(countsTableRound,1,mean),xlim=c(0,3000),ylim=c(0,1000), breaks = 5000)
```

Now we are going to start working with DESeq2!
```{r}
#### Corrects the column names to match between the metadata table and the counts matrix.
colnames(countsTableRound) <-substr(colnames(countsTableRound), start = 1, stop = 4)

#### Create a DESeq object and define the experimental design here with the tilda
dds <- DESeqDataSetFromMatrix(countData = countsTableRound, colData=conds, 
                              design= ~ generation + line)

dim(dds)
```

```{r}
# Filter out genes with too few reads - remove all genes with counts < 15 in more than 75% of 24 samples, so 18)
## suggested by WGCNA on RNAseq FAQ

dds <- dds[rowSums(counts(dds) >= 15) >= 18,]
nrow(dds) 
# How many transcripts have at least 15 reads (a.k.a counts) in 75% of the samples

# Run the DESeq model to test for differential gene expression
dds <- DESeq(dds)

# List the results you've generated
resultsNames(dds)
# Copy the results names 
```
