---
title: "DESeq2_spruce"
output: html_document
date: "2025-11-20"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/projects/eco_genomics_2025/final_project/mydata")
```

Import the libraries that we're likely to need in this session
```{r message=FALSE, warning=FALSE}
library(DESeq2)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(vsn)  
library("pheatmap")
library("vsn")
library(eulerr)
```

Import data:
```{r} 
# Import the counts table
countsTable <- read.csv("counts_matrix.csv", header=TRUE, row.names=1)

countsTableRound <- round(countsTable)
head(countsTableRound)

samples <- colnames(countsTableRound)
removeBLANK <- !grepl("BLANK", samples) # done to remove "BLANK_ACATTA from table"
countsTableRound <- countsTableRound[, removeBLANK]
samples <- colnames(countsTableRound)

# Import the metadata file
metadata_spruce <- read.csv("metadata_spruce.csv", header = TRUE, row.names=1)
```

Explore data distributions:
```{r}
colSums(countsTableRound)
mean(colSums(countsTableRound))
```

Visualizing variation across samples and genes:
```{r}
# variation across samples:
barplot(colSums(countsTableRound), names.arg=colnames(countsTableRound),cex.names=0.5, las=3,ylim=c(0,2500000))
abline(h=mean(colSums(countsTableRound)), col="blue", lwd=2)

# average number of counts per gene:
mean(rowSums(countsTableRound))
median(rowSums(countsTableRound))

# variation across genes:
apply(countsTableRound,2,mean) # 2 in apply, action across columns
apply(countsTableRound,1,mean) # 1 in apply, action across rows
hist(apply(countsTableRound,1,mean),xlim=c(0,14), ylim=c(0,60000),breaks=1000000)
```

Now we are going to start working with DESeq2!
```{r include=FALSE}
#### Create a DESeq object and define the experimental design here with the tilda
dds <- DESeqDataSetFromMatrix(countData = countsTableRound,
                              colData=metadata_spruce, 
                              design= ~ source_climate + day_harvest)
dim(dds)
```

```{r include=FALSE}
# Filter out genes with too few reads - remove all genes with counts < 15 in more than 75% of 24 samples, so 18)
## suggested by WGCNA on RNAseq FAQ

dds <- dds[rowSums(counts(dds) >= 15) >= 18,]
nrow(dds) 
# How many transcripts have at least 15 reads (a.k.a counts) in 75% of the samples

# Run the DESeq model to test for differential gene expression
dds <- DESeq(dds)

# List the results you've generated
resultsNames(dds)
# Copy the results names 
```

# Continuation from Oct 14 class.
```{r include=FALSE}
# The goal of transformation "is to remove the dependence of the variance on the mean, particularly the high variance of the logarithm of count data when the mean is low."

# this gives log2(n + 1)
ntd <- normTransform(dds)
meanSdPlot(assay(ntd))

# Variance stabilizing transformation
vsd <- vst(dds, blind=FALSE)
meanSdPlot(assay(vsd))
```

Make a heatmap of the distance/dissimilarity of the samples:
```{r}
sampleDists <- dist(t(assay(vsd)))

library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$line, vsd$generation, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

# Note any outliers
```

Making a cluster tree to look for outlier samples:
```{r}
sampleTree <- hclust(dist(sampleDists), method="average")
# plot
plot(sampleTree, main="Sample clustering to detect outliers", sub="", xlab="",cex.lab=1.5, cex.axis=1.5, cex.main=2)

# is there a tree that we see on its own? We see two clusters but nothing particularly on its own
```

PCA to visualize global gene expression patterns:
```{r}
# first transform the data for plotting using variance stabilization
vsd <- vst(dds, blind=FALSE)

pcaData <- plotPCA(vsd, intgroup=c("line","generation"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData,"percentVar"))

ggplot(pcaData, aes(PC1, PC2, color=line, shape=generation)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()
```

```{r}
ggplot(pcaData, aes(PC1, PC2)) +
  geom_point(size=5, stroke = 1, aes(fill=line, shape=generation, alpha = generation)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  scale_shape_manual(values=c(21,22,23,24), labels = c("G1", "G2","G3", "G4"))+
  #scale_shape_manual(values=c(21,22,23,24) -> circle, square, diamond, triangle
  scale_fill_manual(values=c('#6699CC',"#CC3333"), labels = c("Control", "Treatment"))+
  scale_alpha_manual(values = c(1.0, 0.8, 0.7, 0.3), guide = "none") +
  theme_bw() +
  theme(legend.position = "none") +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 3))+
  theme(text = element_text(size = 20)) +
  theme(legend.title = element_blank())
```