---
title: "DESeq2_spruce"
output: html_document
date: "2025-11-20"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/projects/eco_genomics_2025/final_project/mydata")
```

Import libraries:
```{r message=FALSE, warning=FALSE}
library(DESeq2)
library(dplyr)
library(tidyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(vsn)  
library("pheatmap")
library("vsn")
library(eulerr)
```

Import data:
```{r} 
# Import the counts table
countsTable <- read.csv("counts_matrix.csv", header=TRUE, row.names=1)

countsTableRound <- round(countsTable)
head(countsTableRound)

samples <- colnames(countsTableRound)
removeBLANK <- !grepl("BLANK", samples) # remove "BLANK_ACATTA from table"
countsTableRound <- countsTableRound[, removeBLANK]
samples <- colnames(countsTableRound)

# Import the metadata file
metadata_spruce <- read.csv("metadata_spruce.csv", header = TRUE, row.names=1)
```

Explore data distributions:
```{r}
colSums(countsTableRound)
mean(colSums(countsTableRound))
```

Visualizing variation across samples and genes:
```{r}
# variation across samples:
barplot(colSums(countsTableRound), names.arg=colnames(countsTableRound),cex.names=0.5, las=3,ylim=c(0,2500000))
abline(h=mean(colSums(countsTableRound)), col="blue", lwd=2)

# average number of counts per gene:
mean(rowSums(countsTableRound))
median(rowSums(countsTableRound))

# variation across genes:
apply(countsTableRound,2,mean) # 2 in apply, action across columns
apply(countsTableRound,1,mean) # 1 in apply, action across rows
hist(apply(countsTableRound,1,mean),xlim=c(0,6), ylim=c(0,60000),breaks=1000000)
```

Now we are going to start working with DESeq2!
```{r}
# subsetting to only include data from day 10
day10_samps = grep("_10_",colnames(countsTableRound))

# creating DESeq objects and defining the experimental design here with the tilda

dds <- DESeqDataSetFromMatrix(countData = countsTableRound[,day10_samps],
                              colData=metadata_spruce[which(metadata_spruce$day_harvest == 10),],
                              design= ~ source_climate + treatment + source_climate:treatment)

dim(dds)

# Filter out genes with too few reads by removing all genes with counts < 15
dds <- dds[rowSums(counts(dds) >= 15) >= 18,]
nrow(dds) 

# Run the DESeq model to test for differential gene expression
dds <- DESeq(dds)

# List the results you've generated
resultsNames(dds)

# Copy the results names 
```

```{r}
# Transformation to remove the dependence of the variance on the mean, particularly the high variance of the logarithm of count data when the mean is low.

# this gives log2(n + 1)
ntd <- normTransform(dds)
meanSdPlot(assay(ntd))

# Variance stabilizing transformation
vsd <- vst(dds, blind=FALSE)
meanSdPlot(assay(vsd))
```

Make a heatmap of the distance/dissimilarity of the samples:
```{r}
sampleDists <- dist(t(assay(vsd)))

library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$source_climate, vsd$treatment, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

# Note any outliers
```

Making a cluster tree to look for outlier samples:
```{r}
sampleTree <- hclust(dist(sampleDists), method = "average")

# plot
plot(sampleTree, main="Sample clustering to detect outliers", sub="", xlab="",cex.lab=1.5, cex.axis=1.5, cex.main=2)
```

PCA to visualize global gene expression patterns:
```{r}
# first transform the data for plotting using variance stabilization
vsd <- vst(dds, blind=FALSE)

pcaData <- plotPCA(vsd, intgroup=c("treatment","source_climate"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData,"percentVar"))

PCAglobal <- ggplot(pcaData, aes(PC1, PC2, color=treatment, shape=source_climate)) +
              geom_point(size=3) +
              xlab(paste0("PC1: ",percentVar[1],"% variance")) +
              ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
              coord_fixed()

PCAglobal

#ggsave("PCAglobal.png", plot = PCAglobal, width = 7, height = 5, units = "in", dpi = 300)
```

```{r}
ggplot(pcaData, aes(PC1, PC2)) +
  geom_point(size=5, stroke = 1, aes(fill = source_climate, shape = treatment, alpha = day_harvest)) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) +
  scale_fill_manual(values=c('#6699CC',"#CC3333"), labels = c("Control", "Treatment"))+
  theme_linedraw() +
  theme(legend.position = "none") +
  theme(panel.border = element_rect(color = "black", fill = NA, size = 3))+
  theme(text = element_text(size = 20)) +
  theme(legend.title = element_blank()) +
  theme_linedraw()
```

There appears to be some clustering...!!!! But let's try specific contrasts just in case.
```{r}
dds$treatment <- relevel(dds$treatment, ref = "Control")
dds$source_climate <- relevel(dds$source_climate, ref = "CoolWet")

dds$group <- factor(paste0(dds$source_climate, dds$treatment))
design(dds) <- ~ group
dds <- DESeq(dds)
dds_new <- resultsNames(dds)
```

Comparing each of the three treatments between CoolWet and HotDry source climates:
```{r}
#Key for labeling treatments for code below: C = control, H = heat, HD = heat; drought

#CoolWet:
coolwetCvsH <- results(dds, contrast=c("group","CoolWetControl","CoolWetHeat"), alpha = 0.05)

coolwetCvsDH <- results(dds, contrast=c("group","CoolWetControl","CoolWetDrought; Heat"), alpha = 0.05)

coolwetHvsDH <- results(dds, contrast=c("group","CoolWetHeat","CoolWetDrought; Heat"), alpha = 0.05)

#HotDry:
hotdryCvsH <- results(dds, contrast=c("group","HotDryControl","HotDryHeat"), alpha = 0.05)

hotdryCvsDH <- results(dds, contrast=c("group","HotDryControl","HotDryDrought; Heat"), alpha = 0.05)

hotdryHvsDH <- results(dds, contrast=c("group","HotDryHeat","HotDryDrought; Heat"), alpha = 0.05)

```

Now we are making an MA plots:

- Question: What is the relationship between LFC and magnitude of expression?
```{r}
#CoolWet:
plotMA(coolwetCvsH, ylim=c(-5,5))
plotMA(coolwetCvsDH, ylim=c(-5,5))
plotMA(coolwetHvsDH, ylim=c(-5,5))

#HotDry:
plotMA(hotdryCvsH, ylim=c(-5,5))
plotMA(hotdryCvsDH, ylim=c(-5,5))
plotMA(hotdryHvsDH, ylim=c(-5,5))
```

Making volcano plots:

- Question: What is the relationship between LFC and significance of DGE?
- First make it into a dataframe and remove NAs, add column to dataframe called "significance" and include adjust parameters showing "up" and "down" regulated genes. Then using ggplot with log fold change on the X axis and adjusted p-value on Y axis. Using the new column of significance to show up and downregulated genes, then giving lines to show log fold change and p-value.

CoolWet across treatment groups:
```{r}
# CoolWet - Control vs. Heat:
coolwetCvsH_res_df <- as.data.frame(coolwetCvsH) # Make a dataframe
coolwetCvsH_res_df <- coolwetCvsH_res_df[!is.na(coolwetCvsH_res_df$padj), ] # Remove NAs
coolwetCvsH_res_df$significance <- "Not Sig" # Add significance column for color mapping
coolwetCvsH_res_df$significance[coolwetCvsH_res_df$padj < 0.05 & coolwetCvsH_res_df$log2FoldChange > 0] <- "Up"
coolwetCvsH_res_df$significance[coolwetCvsH_res_df$padj < 0.05 & coolwetCvsH_res_df$log2FoldChange < 0] <- "Down"

volcano_coolwetCvsH <- ggplot(coolwetCvsH_res_df, aes(x = log2FoldChange, y = -log10(padj), color = significance)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = c("Up" = "blue", "Down" = "red", "Not Sig" = "grey")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +
  labs(
    title = "Volcano Plot: Coolwet - Control vs. Heat",
    x = expression(log[2]~Fold~Change),
    y = expression(-log[10]~Adjusted~p~value)
  ) +
  theme_bw(base_size = 16) +
  theme(legend.title = element_blank())

volcano_coolwetCvsH

### CoolWet - Control vs. Drought; Heat:
coolwetCvsDH_res_df <- as.data.frame(coolwetCvsDH) # Make a dataframe
coolwetCvsDH_res_df <- coolwetCvsDH_res_df[!is.na(coolwetCvsDH_res_df$padj), ] # Remove NAs
coolwetCvsDH_res_df$significance <- "Not Sig" # Add significance column for color mapping
coolwetCvsDH_res_df$significance[coolwetCvsDH_res_df$padj < 0.05 & coolwetCvsDH_res_df$log2FoldChange > 0] <- "Up"
coolwetCvsDH_res_df$significance[coolwetCvsDH_res_df$padj < 0.05 & coolwetCvsDH_res_df$log2FoldChange < 0] <- "Down"

volcano_coolwetCvsDH <- ggplot(coolwetCvsDH_res_df, aes(x = log2FoldChange, y = -log10(padj), color = significance)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = c("Up" = "blue", "Down" = "red", "Not Sig" = "grey")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +
  labs(
    title = "Volcano Plot: Coolwet - Control vs. Drought; Heat",
    x = expression(log[2]~Fold~Change),
    y = expression(-log[10]~Adjusted~p~value)
  ) +
  theme_bw(base_size = 16) +
  theme(legend.title = element_blank())

volcano_coolwetCvsDH

# CoolWet - Heat vs. Drought; Heat:
coolwetHvsDH_res_df <- as.data.frame(coolwetHvsDH) # Make a dataframe
coolwetHvsDH_res_df <- coolwetHvsDH_res_df[!is.na(coolwetHvsDH_res_df$padj), ] # Remove NAs
coolwetHvsDH_res_df$significance <- "Not Sig" # Add significance column for color mapping
coolwetHvsDH_res_df$significance[coolwetHvsDH_res_df$padj < 0.05 & coolwetHvsDH_res_df$log2FoldChange > 0] <- "Up"
coolwetHvsDH_res_df$significance[coolwetHvsDH_res_df$padj < 0.05 & coolwetHvsDH_res_df$log2FoldChange < 0] <- "Down"

volcano_coolwetHvsDH <- ggplot(coolwetHvsDH_res_df, aes(x = log2FoldChange, y = -log10(padj), color = significance)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = c("Up" = "blue", "Down" = "red", "Not Sig" = "grey")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +
  labs(
    title = "Volcano Plot: CoolWet - Heat vs. Drought; Heat",
    x = expression(log[2]~Fold~Change),
    y = expression(-log[10]~Adjusted~p~value)
  ) +
  theme_bw(base_size = 16) +
  theme(legend.title = element_blank())

volcano_coolwetHvsDH
```

HotDry across treatment groups:
```{r}
# HotDry - Control vs. Heat:
hotdryCvsH_res_df <- as.data.frame(hotdryCvsH) # Make a dataframe
hotdryCvsH_res_df <- hotdryCvsH_res_df[!is.na(hotdryCvsH_res_df$padj), ] # Remove NAs
hotdryCvsH_res_df$significance <- "Not Sig" # Add significance column for color mapping
hotdryCvsH_res_df$significance[hotdryCvsH_res_df$padj < 0.05 & hotdryCvsH_res_df$log2FoldChange > 0] <- "Up"
hotdryCvsH_res_df$significance[hotdryCvsH_res_df$padj < 0.05 & hotdryCvsH_res_df$log2FoldChange < 0] <- "Down"

volcano_hotdryCvsH <- ggplot(hotdryCvsH_res_df, aes(x = log2FoldChange, y = -log10(padj), color = significance)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = c("Up" = "blue", "Down" = "red", "Not Sig" = "grey")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +
  labs(
    title = "Volcano Plot: HotDry - Control vs. Heat",
    x = expression(log[2]~Fold~Change),
    y = expression(-log[10]~Adjusted~p~value)
  ) +
  theme_bw(base_size = 16) +
  theme(legend.title = element_blank())

volcano_hotdryCvsH

# HotDry - Control vs. Drought; Heat:
hotdryCvsDH_res_df <- as.data.frame(hotdryCvsDH) # Make a dataframe
hotdryCvsDH_res_df <- hotdryCvsDH_res_df[!is.na(hotdryCvsDH_res_df$padj), ] # Remove NAs
hotdryCvsDH_res_df$significance <- "Not Sig" # Add significance column for color mapping
hotdryCvsDH_res_df$significance[hotdryCvsDH_res_df$padj < 0.05 & hotdryCvsDH_res_df$log2FoldChange > 0] <- "Up"
hotdryCvsDH_res_df$significance[hotdryCvsDH_res_df$padj < 0.05 & hotdryCvsDH_res_df$log2FoldChange < 0] <- "Down"

volcano_hotdryCvsDH <- ggplot(hotdryCvsDH_res_df, aes(x = log2FoldChange, y = -log10(padj), color = significance)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = c("Up" = "blue", "Down" = "red", "Not Sig" = "grey")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +
  labs(
    title = "Volcano Plot: HotDry - Control vs. Drought; Heat",
    x = expression(log[2]~Fold~Change),
    y = expression(-log[10]~Adjusted~p~value)
  ) +
  theme_bw(base_size = 16) +
  theme(legend.title = element_blank())

volcano_hotdryCvsDH

# HotDry - Heat vs. Drought; Heat:
hotdryHvsDH_res_df <- as.data.frame(hotdryHvsDH) # Make a dataframe
hotdryHvsDH_res_df <- hotdryHvsDH_res_df[!is.na(hotdryHvsDH_res_df$padj), ] # Remove NAs
hotdryHvsDH_res_df$significance <- "Not Sig" # Add significance column for color mapping
hotdryHvsDH_res_df$significance[hotdryHvsDH_res_df$padj < 0.05 & hotdryHvsDH_res_df$log2FoldChange > 0] <- "Up"
hotdryHvsDH_res_df$significance[hotdryHvsDH_res_df$padj < 0.05 & hotdryHvsDH_res_df$log2FoldChange < 0] <- "Down"

volcano_hotdryHvsDH <- ggplot(hotdryHvsDH_res_df, aes(x = log2FoldChange, y = -log10(padj), color = significance)) +
  geom_point(alpha = 0.7, size = 2) +
  scale_color_manual(values = c("Up" = "blue", "Down" = "red", "Not Sig" = "grey")) +
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "black") +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "black") +
  labs(
    title = "Volcano Plot: HotDry - Heat vs. Drought; Heat",
    x = expression(log[2]~Fold~Change),
    y = expression(-log[10]~Adjusted~p~value)
  ) +
  theme_bw(base_size = 16) +
  theme(legend.title = element_blank())

volcano_hotdryHvsDH
```

Heatmap #1 - top 20 genes sorted by p-value for both source climates across treatment groups:
- Question: For DEGs, what does gene expression variation look like across individual samples among treatment groups?

```{r}
# CoolWet - Control vs. Heat:
cwCvsH_topgenes <- head(rownames(coolwetCvsH_res_df),10)
cwCvsH_mat <- assay(vsd)[cwCvsH_topgenes,]
cwCvsH_mat <- cwCvsH_mat - rowMeans(cwCvsH_mat)
cwCvsH_df <- as.data.frame(colData(dds)[,c("source_climate","treatment")])
pheatmap(cwCvsH_mat, annotation_col=cwCvsH_df)
pheatmap(cwCvsH_mat, annotation_col=cwCvsH_df, cluster_cols = F)

# CoolWet - Control vs. Drought; Heat:
cwCvsDH_topgenes <- head(rownames(coolwetCvsDH_res_df),10)
cwCvsDH_mat <- assay(vsd)[cwCvsDH_topgenes,]
cwCvsDH_mat <- cwCvsDH_mat - rowMeans(cwCvsDH_mat)
cwCvsDH_df <- as.data.frame(colData(dds)[,c("source_climate","treatment")])
pheatmap(cwCvsDH_mat, annotation_col=cwCvsDH_df)
pheatmap(cwCvsDH_mat, annotation_col=cwCvsDH_df, cluster_cols = F)

# CoolWet - Heat vs. Drought; Heat:
cwHvsDH_topgenes <- head(rownames(coolwetHvsDH_res_df),10)
cwHvsDH_mat <- assay(vsd)[cwHvsDH_topgenes,]
cwHvsDH_mat <- cwHvsDH_mat - rowMeans(cwHvsDH_mat)
cwHvsDH_df <- as.data.frame(colData(dds)[,c("source_climate","treatment")])
pheatmap(cwHvsDH_mat, annotation_col=cwHvsDH_df)
pheatmap(cwHvsDH_mat, annotation_col=cwHvsDH_df, cluster_cols = F)

# HotDry - Control vs. Heat:
hdCvsH_topgenes <- head(rownames(hotdryCvsH_res_df),10)
hdCvsH_mat <- assay(vsd)[hdCvsH_topgenes,]
hdCvsH_mat <- hdCvsH_mat - rowMeans(hdCvsH_mat)
hdCvsH_df <- as.data.frame(colData(dds)[,c("source_climate","treatment")])
pheatmap(hdCvsH_mat, annotation_col=hdCvsH_df)
pheatmap(hdCvsH_mat, annotation_col=hdCvsH_df, cluster_cols = F)

# HotDry - Control vs. Drought; Heat:
hdCvsDH_topgenes <- head(rownames(hotdryCvsDH_res_df),10)
hdCvsDH_mat <- assay(vsd)[hdCvsDH_topgenes,]
hdCvsDH_mat <- hdCvsDH_mat - rowMeans(hdCvsDH_mat)
hdCvsDH_df <- as.data.frame(colData(dds)[,c("source_climate","treatment")])
pheatmap(hdCvsDH_mat, annotation_col=hdCvsDH_df)
pheatmap(hdCvsDH_mat, annotation_col=hdCvsDH_df, cluster_cols = F)

# HotDry - Heat vs. Drought; Heat:
hdHvsDH_topgenes <- head(rownames(hotdryHvsDH_res_df),10)
hdHvsDH_mat <- assay(vsd)[hdHvsDH_topgenes,]
hdHvsDH_mat <- cwHvsDH_mat - rowMeans(hdHvsDH_mat)
hdHvsDH_df <- as.data.frame(colData(dds)[,c("source_climate","treatment")])
pheatmap(hdHvsDH_mat, annotation_col=hdHvsDH_df)
pheatmap(hdHvsDH_mat, annotation_col=hdHvsDH_df, cluster_cols = F)
```

Let’s make plots to explore differentially expressed genes between treatment groups across source climates:

```{r}
# First step: Identify the DEGs in the contrast for CoolWet - Control vs. Treatment group for each source climate:

#Coolwet - Control vs. Heat
res_coolwetCvsH <- results(dds, contrast=c("group","CoolWetControl","CoolWetHeat"), alpha = 0.05)
res_coolwetCvsH <- res_coolwetCvsH[order(res_coolwetCvsH$padj),] #sort
head(res_coolwetCvsH)
summary(res_coolwetCvsH)

res_coolwetCvsH <- res_coolwetCvsH[!is.na(res_coolwetCvsH$padj),]
degs_coolwetCvsH <- row.names(res_coolwetCvsH[res_coolwetCvsH$padj < 0.05,])

#Coolwet - Control vs. Drought; Heat
res_coolwetCvsDH <- results(dds, contrast=c("group","CoolWetControl","CoolWetDrought; Heat"), alpha =0.05)
res_coolwetCvsDH <- res_coolwetCvsDH[order(res_coolwetCvsDH$padj),] #sort
head(res_coolwetCvsDH)
summary(res_coolwetCvsDH)

res_coolwetCvsDH <- res_coolwetCvsDH[!is.na(res_coolwetCvsDH$padj),]
degs_coolwetCvsDH <- row.names(res_coolwetCvsDH[res_coolwetCvsDH$padj < 0.05,])

#Coolwet - Heat vs. Drought; Heat
res_coolwetHvsDH <- results(dds, contrast=c("group","CoolWetHeat","CoolWetDrought; Heat"), alpha = 0.05)
res_coolwetHvsDH <- res_coolwetHvsDH[order(res_coolwetHvsDH$padj),] #sort
head(res_coolwetHvsDH)
summary(res_coolwetHvsDH)

res_coolwetHvsDH <- res_coolwetHvsDH[!is.na(res_coolwetHvsDH$padj),]
degs_coolwetHvsDH <- row.names(res_coolwetHvsDH[res_coolwetHvsDH$padj < 0.05,])
```

```{r}
#HotDry - Control vs. Heat
res_hotdryCvsH <- results(dds, contrast=c("group","HotDryControl","HotDryHeat"), alpha = 0.05)
res_hotdryCvsH <- res_hotdryCvsH[order(res_hotdryCvsH$padj),] #sort
head(res_hotdryCvsH)
summary(res_hotdryCvsH)

res_hotdryCvsH <- res_hotdryCvsH[!is.na(res_hotdryCvsH$padj),]
degs_res_hotdryCvsH <- row.names(res_hotdryCvsH[res_hotdryCvsH$padj < 0.05,])

#HotDry - Control vs. Drought; Heat
res_hotdryCvsDH <- results(dds, contrast=c("group","HotDryControl","HotDryDrought; Heat"), alpha =0.05)
res_hotdryCvsDH <- res_hotdryCvsDH[order(res_hotdryCvsDH$padj),] #sort
head(res_hotdryCvsDH)
summary(res_hotdryCvsDH)

res_hotdryCvsDH <- res_hotdryCvsDH[!is.na(res_hotdryCvsDH$padj),]
degs_hotdryCvsDH <- row.names(res_hotdryCvsDH[res_hotdryCvsDH$padj < 0.05,])

#HotDry - Heat vs. Drought; Heat
res_hotdryHvsDH <- results(dds, contrast=c("group","HotDryHeat","HotDryDrought; Heat"), alpha = 0.05)
res_hotdryHvsDH <- res_hotdryHvsDH[order(res_hotdryHvsDH$padj),] #sort
head(res_hotdryHvsDH)
summary(res_hotdryHvsDH)

res_hotdryHvsDH <- res_hotdryHvsDH[!is.na(res_hotdryHvsDH$padj),]
degs_hotdryHvsDH <- row.names(res_hotdryHvsDH[res_hotdryHvsDH$padj < 0.05,])
```

Make a heatmap of treatment significant genes and how they change across source_climate:

- Question: How does gene expression change across treatments compared to source climate?

- Start by identifying significant genes, then make a dataframe with log fold change for all of the significant genes from treatment types (we'll call this log fold change matrix), then remove rows with any same values, and use a heatmap to plot the dataframe and scale it by rows and then by color. Cluster by rows and colors by equal distance.

LFG of CoolWet across treatments:
```{r}
# CoolWet:
cw_sig_genes <- rownames(coolwetCvsDH[which(coolwetCvsDH$padj < 0.05 & !is.na(coolwetCvsDH$padj)), ])

# Combine into a data frame
cw_lfc_mat <- data.frame(
  CvsH = coolwetCvsH[cw_sig_genes, "log2FoldChange"],
  CvDH = coolwetCvsDH[cw_sig_genes, "log2FoldChange"],
  HvDH = coolwetHvsDH[cw_sig_genes, "log2FoldChange"]
)

# Remove any rows with missing values (some genes may not be in all result tables)
cw_lfc_mat <- na.omit(cw_lfc_mat)

pheatmap(cw_lfc_mat,
         scale = "row",             # scales each gene’s LFCs across generations
         color = colorRampPalette(c("blue", "white", "red"))(50),
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         show_rownames = FALSE,
         main = "LFC of Sig Genes from CoolWet source climate across treatments")
```

LFG of HotDry across treatments:
```{r}
hd_sig_genes <- rownames(hotdryCvsDH[which(hotdryCvsDH$padj < 0.05 & !is.na(hotdryCvsDH$padj)), ])

# Combine into a data frame
hd_lfc_mat <- data.frame(
  CvsH = hotdryCvsH[hd_sig_genes, "log2FoldChange"],
  CvDH = hotdryCvsDH[hd_sig_genes, "log2FoldChange"],
  HvDH = hotdryHvsDH[hd_sig_genes, "log2FoldChange"]
)

# Remove any rows with missing values (some genes may not be in all result tables)
hd_lfc_mat <- na.omit(hd_lfc_mat)

pheatmap(hd_lfc_mat,
         scale = "row",             # scales each gene’s LFCs across generations
         color = colorRampPalette(c("blue", "white", "red"))(50),
         clustering_distance_rows = "euclidean",
         clustering_distance_cols = "euclidean",
         clustering_method = "complete",
         show_rownames = FALSE,
         main = "LFC of Sig Genes from HotDry source climate across treatments")
```

## DESeq2 To TopGO:

Import libraries and data:
```{r}
library(topGO)
library(GO.db)

z <- read.table("/gpfs1/cl/ecogen/pbio6800/PopulationGenomics/ref_genome/Pmariana/annotation/GO_annotations/Pmariana_GO_annotations_sorted_join.txt", header=F, fill=T)
PmGOannot <- z %>% mutate(V7=ifelse(V7 %in% "", V2,V7))
colnames(PmGOannot) <- c("gene", "start", "stop", "geneID", "strand", "parent", "marker")

PmGOannot$geneID <- gsub('ID=', '', PmGOannot$geneID)
```

** Keep on getting this: "Warning :No enrichment can pe performed - there are no feasible GO terms!"

1. Functional enrichment for CoolWet - Control vs. Heat
```{r}
# Convert to data frame
coolwetCvsH_df <- as.data.frame(coolwetCvsH)

# Add geneID column from rownames
coolwetCvsH_df$geneID <- rownames(coolwetCvsH)
coolwetCvsH_df <- coolwetCvsH_df[, c("geneID", setdiff(names(coolwetCvsH_df), "geneID"))]

# Read GO terms file
spruce_go <- read.delim("Pmariana_GOannot.tsv", header = TRUE) %>%
  slice(-c(1,2)) %>%  # remove header rows if present
  na.omit()  # remove missing GO annotations

coolwetCvsH_merge <- merge(coolwetCvsH_df, PmGOannot, by="geneID")

# Identify DEGs that have GO annotations
cwCvsH_GO <- coolwetCvsH_merge[coolwetCvsH_merge$geneID %in% spruce_go$geneID, ]

# Keep only significant DEGs 
cwCvsH_GO_sig <- subset(cwCvsH_GO, padj < 0.05 & abs(log2FoldChange) > 1)

# Merge with GO annotations
cwCvsH_stat <- merge(cwCvsH_GO_sig, cwCvsH_GO, by = "geneID")

# Create gene2GO list which is used in the topgo function 
cwCvsH_gene2GO <- spruce_go %>%
  filter(!is.na(GO) & GO != "") %>%
  group_by(geneID) %>%
  summarise(GO = list(unique(GO)), .groups = "drop") %>%
  with(setNames(GO, geneID))

# Split semicolon-separated GO terms
cwCvsH_gene2GO <- lapply(cwCvsH_gene2GO, function(x) unlist(strsplit(x, ";")))

# Create geneList for TopGO: 1 = DEG, 0 = not DEG
cwCvsH_allgenes <- names(cwCvsH_gene2GO)
cwCvsH_geneList <- setNames(as.integer(cwCvsH_allgenes %in% cwCvsH_stat$gene), cwCvsH_allgenes)

#Build TopGO pipeline 
cwCvsH_GOdata <- new("topGOdata",
              description = "CoolWet: Control vs. Heat DEGs GO enrichment",
              ontology = "BP",
              allGenes = cwCvsH_geneList,
              geneSel = function(x) x == 1,
              nodeSize = 5,
              annot = annFUN.gene2GO,
              gene2GO = cwCvsH_gene2GO)

#test for enrichment 
cwCvsH_result <- runTest(cwCvsH_GOdata, algorithm = "parentChild", statistic = "fisher")

```

2. Functional enrichment for CoolWet - Control vs. Drought; Heat
```{r}
# Convert to data frame
# Convert to data frame
coolwetCvsDH_df <- as.data.frame(coolwetCvsDH)

# Add geneID column from rownames
coolwetCvsDH_df$geneID <- rownames(coolwetCvsDH)
coolwetCvsDH_df <- coolwetCvsDH_df[, c("geneID", setdiff(names(coolwetCvsDH_df), "geneID"))]

# Read GO terms file
spruce_go <- read.delim("Pmariana_GOannot.tsv", header = TRUE) %>%
  slice(-c(1,2)) %>%  # remove header rows if present
  na.omit()  # remove missing GO annotations

coolwetCvsDH_merge <- merge(coolwetCvsDH_df, PmGOannot, by="geneID")

# Identify DEGs that have GO annotations
cwCvsDH_GO <- coolwetCvsDH_merge[coolwetCvsDH_merge$gene %in% spruce_go$gene, ]

# Keep only significant DEGs 
cwCvsDH_GO_sig <- subset(cwCvsDH_GO, padj < 0.05 & abs(log2FoldChange) > 1)

# Merge with GO annotations
cwCvsDH_stat <- merge(cwCvsDH_GO_sig, cwCvsDH_GO, by = "gene")

# Create gene2GO list which is used in the topgo function 
cwCvsDH_gene2GO <- spruce_go %>%
  filter(!is.na(GO) & GO != "") %>%
  group_by(geneID) %>%
  summarise(GO = list(unique(GO)), .groups = "drop") %>%
  with(setNames(GO, geneID))

# Split semicolon-separated GO terms
cwCvsDH_gene2GO <- lapply(cwCvsDH_gene2GO, function(x) unlist(strsplit(x, ";")))

# Create geneList for TopGO: 1 = DEG, 0 = not DEG
cwCvsDH_allgenes <- names(cwCvsDH_gene2GO)
cwCvsDH_geneList <- setNames(as.integer(cwCvsDH_allgenes %in% cwCvsDH_stat$gene), cwCvsDH_allgenes)

#Build TopGO pipeline 
cwCvsDH_GOdata <- new("topGOdata",
              description = "CoolWet: Control vs. Heat DEGs GO enrichment",
              ontology = "BP",
              allGenes = cwCvsDH_geneList,
              geneSel = function(x) x == 1,
              nodeSize = 5,
              annot = annFUN.gene2GO,
              gene2GO = cwCvsDH_gene2GO)

#test for enrichment 
cwCvsDH_result <- runTest(cwCvsDH_GOdata, algorithm = "parentChild", statistic = "fisher")
```